File: sol/155.cpp
#include <stack>

struct MinStack {
    stack<int> data, order;
    MinStack(): data(), order() {}
    
    void push(int val) {
        this->data.push(val);
        if (this->order.empty() || this->order.top() >= val){
            this->order.push(val);
        }
    }
    
    inline void pop() {
        if (this->order.top() == this->data.top()){
            this->order.pop();
        }
        this->data.pop();
    }
    
    inline int top(void) const { return this->data.top(); }
    
    inline int getMin(void) const { return this->order.top(); }
};

File: sol/155/155.py
class MinStack:
    def __init__(self):
        self.data: list[int] = []
        self.order: list[int] = []

    def push(self, val: int) -> None:
        self.data.append(val)
        if len(self.order) == 0 or self.order[-1] >= val:
            self.order.append(val)
        else:
            self.order.append(self.order[-1])

    def pop(self) -> None:
        self.order.pop()
        return self.data.pop()

    def top(self) -> int:
        return self.data[-1]

    def getMin(self) -> int:
        return self.order[-1]

